### PR TITLE
feat: make hightlighting linear

### PR DESCRIPTION
**Pull request by matklad**
In https://youtu.be/qvIZZf5dmTE, we've noticed that AstIdMap does a
linear lookup when going from SyntaxNode to Id. This leads to
accidentally quadratic overall performance. Replace linear lookup with a
O(1) hashmap lookup.

Future work: don't duplicate `SyntaxNodePtr` in `AstIdMap` and switch to
"call site dependency injection" style storage (eg, store a
`HashSet<ErasedFileAstId>`).

See the explanation of the work here on YouTube :-)

As you can see from then benchmark results, this doesn't actually make analysis stats fastre. I am a bit mystified as to why this is happening to be honest. 

Baseline
```
Database loaded:     598.40ms, 304minstr, 118mb (metadata 390.57ms, 21minstr, 841kb; build 111.31ms, 8764kinstr, -214kb)
  crates: 39, mods: 824, decls: 18647, fns: 13910
Item Collection:     9.70s, 75ginstr, 377mb
  exprs: 382426, ??ty: 387 (0%), ?ty: 285 (0%), !ty: 145
Inference:           43.16s, 342ginstr, 641mb
Total:               52.86s, 417ginstr, 1018mb
```

This PR:
```
Database loaded:     626.34ms, 304minstr, 118mb (metadata 416.26ms, 21minstr, 841kb; build 113.67ms, 8750kinstr, -209kb)
  crates: 39, mods: 824, decls: 18647, fns: 13910
Item Collection:     10.16s, 75ginstr, 389mb
  exprs: 382426, ??ty: 387 (0%), ?ty: 285 (0%), !ty: 145
Inference:           44.51s, 342ginstr, 644mb
Total:               54.67s, 417ginstr, 1034mb
```

I think we probably should merge the first commit here, but not the second.

### CODE REVIEW

**Comment by matklad** [comment]
https://youtu.be/wvEgymUm7cY <- me not understanding what's going on here :-)

**Comment by flodiebold** [comment]
Even if it doesn't improve performance, you'd expect _some_ kind of change in the instruction counts, wouldn't you? :confused:

**Comment by matklad** [comment]
I've more carefully re-did laziness checking, and for analysis stats it seems that although *some* reverse maps are indeed not queried, this is not the common case. I've also gave a shot at using more efficient hash map: 

```rust
// better hash map
Database loaded:     593.14ms, 306minstr, 119mb (metadata 388.94ms, 21minstr, 882kb; build 112.76ms, 8732kinstr, -252kb)
  crates: 39, mods: 827, decls: 18716, fns: 13959
Item Collection:     9.66s, 74ginstr, 382mb
  exprs: 384097, ??ty: 387 (0%), ?ty: 285 (0%), !ty: 144
Inference:           43.54s, 340ginstr, 644mb
Total:               53.19s, 414ginstr, 1027mb

// optimization from last week
Database loaded:     596.02ms, 306minstr, 119mb (metadata 388.88ms, 21minstr, 882kb; build 111.12ms, 8743kinstr, -232kb)
  crates: 39, mods: 827, decls: 18716, fns: 13959
Item Collection:     9.57s, 74ginstr, 390mb
  exprs: 384097, ??ty: 387 (0%), ?ty: 285 (0%), !ty: 144
Inference:           42.71s, 340ginstr, 645mb
Total:               52.27s, 414ginstr, 1036mb

// baseline
Database loaded:     604.75ms, 306minstr, 119mb (metadata 395.82ms, 21minstr, 882kb; build 112.13ms, 8811kinstr, -252kb)
  crates: 39, mods: 827, decls: 18716, fns: 13959
Item Collection:     9.81s, 74ginstr, 378mb
  exprs: 384097, ??ty: 387 (0%), ?ty: 285 (0%), !ty: 144
Inference:           42.39s, 340ginstr, 642mb
Total:               52.20s, 414ginstr, 1020mb
```

It still uses more memory, but not as much memory as the original version.  I think it makes sense to ship this -- more memory is unfortunate, but I feel not having O(N^2) paths is more so. 

While working on it, I also got the impression that maybe we *don't* need to have AstIdMap at all? It seems that it should be possible to move ItemTree into hir_expand, and than make `AstIdMap` data just a  part of that.

**Comment by lnicola** [comment]
@matklad some trailing whitespace there

File: crates/hir_expand/src/ast_id_map.rs [hunk 1]
```diff
@@ -117,6 +145,12 @@ impl AstIdMap {
     }
 }
 
+fn hash_ptr(ptr: &SyntaxNodePtr) -> u64 {
+    let mut hasher = BuildHasherDefault::<FxHasher>::default().build_hasher();
```
Comment (coolreader18) [inline crates/hir_expand/src/ast_id_map.rs:149]: Just looking at this from the weekly feature report, but isn't this line equivalent to `let mut hasher = FxHasher::default();`?

Comment (matklad) [reply crates/hir_expand/src/ast_id_map.rs:149]: Probably! I never really wrapped my head around build hashed thing :-)

**Pull request merged**